jwt/keyset.go

```diff
@@ -0,0 +1,239 @@
+package jwt
+
+import (
+	"context"
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/rsa"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/json"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"mime"
+	"net/http"
+	"strings"
+
+	"github.com/coreos/go-oidc"
+	"github.com/hashicorp/go-cleanhttp"
+	"golang.org/x/oauth2"
+	"gopkg.in/square/go-jose.v2/jwt"
+)
+
+// KeySet represents a set of keys that can be used to verify the signatures of JWTs.
+// A KeySet is expected to be backed by a set of local or remote keys.
+type KeySet interface {
+
+	// VerifySignature parses the given JWT, verifies its signature, and returns the claims in its payload.
+	// The given JWT must be of the JWS compact serialization form.
+	VerifySignature(ctx context.Context, token string) (claims map[string]interface{}, err error)
+}
+
+// jsonWebKeySet verifies JWT signatures using keys obtained from a JWKS URL.
+type jsonWebKeySet struct {
+	remoteJWKS oidc.KeySet
+}
+
+// staticKeySet verifies JWT signatures using local public keys.
+type staticKeySet struct {
+	publicKeys []crypto.PublicKey
+}
+
+// NewOIDCDiscoveryKeySet returns a KeySet that verifies JWT signatures using keys from the
+// JSON Web Key Set (JWKS) published in the discovery document at the given issuer URL.
+// The client used to obtain the remote keys will verify server certificates using the root
+// certificates provided by issuerCAPEM. If issuerCAPEM is not provided, system certificates
+// are used.
+func NewOIDCDiscoveryKeySet(ctx context.Context, issuer string, issuerCAPEM string) (KeySet, error) {
+	if issuer == "" {
+		return nil, errors.New("issuer must not be empty")
+	}
+
+	// Configure an http client with the given certificates
+	caCtx, err := createCAContext(ctx, issuerCAPEM)
+	if err != nil {
+		return nil, err
+	}
+	client := http.DefaultClient
+	if c, ok := caCtx.Value(oauth2.HTTPClient).(*http.Client); ok {
+		client = c
+	}
+
+	// Create and send the http request for the OIDC discovery document
+	wellKnown := strings.TrimSuffix(issuer, "/") + "/.well-known/openid-configuration"
+	req, err := http.NewRequest(http.MethodGet, wellKnown, nil)
+	if err != nil {
+		return nil, err
+	}
+	resp, err := client.Do(req.WithContext(caCtx))
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	// Read the response body and status code
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("unable to read response body: %v", err)
+	}
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("%s: %s", resp.Status, body)
+	}
+
+	// Unmarshal the response body to obtain the issuer and JWKS URL
+	type providerJSON struct {
```

Comment:
A type seems superfluous here. Maybe just:
```
	var p struct {
		Issuer  string `json:"issuer"`
		JWKSURL string `json:"jwks_uri"`
	}
```

Reply:
Ah, I forgot about the anonymous struct. Thanks! Done in 0792e44.